### PR TITLE
[iam,permhandler] Add Permission Handler with unit tests

### PR DESCRIPTION
**Pull request by mlohvynenko**

### CODE REVIEW

File: src/iam/permhandler/permhandler.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    Tie(uuidSecret, err) = GetSecretForInstance(instanceIdent);
+    if (err.IsNone()) {
+        LOG_DBG() << "Instance " << instanceIdent << " has been already registered";
+
+        return {uuidSecret, err};
+    }
+
+    uuidSecret = GenerateSecret();
+    err        = AddSecret(uuidSecret, instanceIdent, instancePermissions);
+
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:36]: Remove empty line:
```c++
err = ...
if (!err.IsNone) {
...
}
```
it is common template

File: src/iam/permhandler/permhandler.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    StaticString<uuid::cUUIDStrLen> uuidSecret;
+
+    Tie(uuidSecret, err) = GetSecretForInstance(instanceIdent);
+    if (err.IsNone()) {
+        LOG_DBG() << "Instance " << instanceIdent << " has been already registered";
+
+        return {uuidSecret, err};
+    }
+
+    uuidSecret = GenerateSecret();
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:34]: add empty line

File: src/iam/permhandler/permhandler.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    return {uuidSecret, err};
+}
+
+Error PermHandler::UnregisterInstance(const InstanceIdent& instanceIdent)
+{
+    LockGuard lock(mMutex);
+
+    auto err
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:48]: `auto result = ...`

File: src/iam/permhandler/permhandler.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        = mInstancesPerms.Find([&instanceIdent](const auto& item) { return item.mInstanceIdent == instanceIdent; });
+    if (!err.mError.IsNone()) {
+        LOG_WRN() << "Instance " << instanceIdent << " not registered";
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:51]: Log pattern and add upper log:
```c++
LOG_DBG() << "Unregister instance: instance = " << instanceIdent;
```

File: src/iam/permhandler/permhandler.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    return mInstancesPerms.Remove(err.mValue).mError;
+}
+
+Error PermHandler::GetPermissions(InstanceIdent& instanceIdent, Array<PermKeyValue>& servicePermissions,
+    const String& secretUUID, const String& funcServerID)
+{
+    LockGuard lock(mMutex);
+
+    auto err = mInstancesPerms.Find([&secretUUID](const auto& item) { return secretUUID == item.mSecretUUID; });
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:64]: `auto result = ...`

File: src/iam/permhandler/permhandler.cpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    if (!err.mError.IsNone()) {
+        LOG_ERR() << "secret not found";
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:66]: No need for error log

File: src/iam/permhandler/permhandler.cpp [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+Error PermHandler::UnregisterInstance(const InstanceIdent& instanceIdent)
+{
+    LockGuard lock(mMutex);
+
+    auto err
+        = mInstancesPerms.Find([&instanceIdent](const auto& item) { return item.mInstanceIdent == instanceIdent; });
+    if (!err.mError.IsNone()) {
+        LOG_WRN() << "Instance " << instanceIdent << " not registered";
+
+        return err.mError;
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:53]: wrap error

File: src/iam/permhandler/permhandler.cpp [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+Error PermHandler::GetPermissions(InstanceIdent& instanceIdent, Array<PermKeyValue>& servicePermissions,
+    const String& secretUUID, const String& funcServerID)
+{
+    LockGuard lock(mMutex);
+
+    auto err = mInstancesPerms.Find([&secretUUID](const auto& item) { return secretUUID == item.mSecretUUID; });
+    if (!err.mError.IsNone()) {
+        LOG_ERR() << "secret not found";
+
+        return err.mError;
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:68]: wrap error

File: src/iam/permhandler/permhandler.cpp [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+            servicePermissions = it.mPermissions;
+
+            return ErrorEnum::eNone;
+        }
+    }
+
+    LOG_ERR() << "permissions for functional server " << funcServerID << " not found";
+
+    return ErrorEnum::eNotFound;
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:87]: Wrap log

File: src/iam/permhandler/permhandler.cpp [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+                return ErrorEnum::eNoMemory;
+            }
+
+            servicePermissions = it.mPermissions;
+
+            return ErrorEnum::eNone;
+        }
+    }
+
+    LOG_ERR() << "permissions for functional server " << funcServerID << " not found";
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:85]: This log is not necessary

File: src/iam/permhandler/permhandler.cpp [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+ * Private
+ **********************************************************************************************************************/
+
+Error PermHandler::AddSecret(const String& secretUUID, const InstanceIdent& instanceIdent,
+    const Array<FunctionalServicePermissions>& instancePermissions)
+{
+    if (mInstancesPerms.Size() == mInstancesPerms.MaxSize()) {
+        LOG_ERR() << "Max number of secrets stored has been reached";
+
+        return ErrorEnum::eNoMemory;
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:100]: Wrap log

File: src/iam/permhandler/permhandler.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+Error PermHandler::AddSecret(const String& secretUUID, const InstanceIdent& instanceIdent,
+    const Array<FunctionalServicePermissions>& instancePermissions)
+{
+    if (mInstancesPerms.Size() == mInstancesPerms.MaxSize()) {
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:97]: Don't think we need to check it as PushBack does it for you

File: src/iam/permhandler/permhandler.cpp [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+StaticString<uuid::cUUIDStrLen> PermHandler::GenerateSecret()
+{
+    const auto uuid = uuid::CreateUUID();
+
+    return uuid::UUIDToString(uuid);
+}
+
+RetWithError<StaticString<uuid::cUUIDStrLen>> PermHandler::GetSecretForInstance(const InstanceIdent& instanceIdent)
+{
+    const auto err
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:115]: `auto result`

File: src/iam/permhandler/permhandler.cpp [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        = mInstancesPerms.Find([&instanceIdent](const auto& elem) { return instanceIdent == elem.mInstanceIdent; });
+
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:117]: remove empty line

File: src/iam/permhandler/permhandler.cpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    if (!err.mError.IsNone()) {
+        return {"", ErrorEnum::eNotFound};
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:119]: wrap error and return `result.mError`

File: src/iam/permhandler/permhandler.cpp [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+        return ErrorEnum::eNoMemory;
+    }
+
+    return mInstancesPerms.PushBack(InstancePermissions {secretUUID, instanceIdent, instancePermissions});
+}
+
+StaticString<uuid::cUUIDStrLen> PermHandler::GenerateSecret()
+{
+    const auto uuid = uuid::CreateUUID();
```
Comment (al1img) [inline src/iam/permhandler/permhandler.cpp:108]: Our UUID generator unfortunately doesn't guarantee unique values. So we need to keep check unique and retry mechanism. Similar we have in go implementation.

File: tests/iam/permhandler/permhandler_test.cpp [hunk 17]
```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "aos/common/tools/buffer.hpp"
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:8]: Group and sourt logs:
system
3rd parties
local

File: tests/iam/permhandler/permhandler_test.cpp [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+
+/***********************************************************************************************************************
+ * Suite
+ **********************************************************************************************************************/
+
+class PermHandlerTest : public Test {
+protected:
+    void SetUp() override
+    {
+        Log::SetCallback([](LogModule module, LogLevel level, const String& message) {
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:27]: We have init logs function in test utils.

File: tests/iam/permhandler/permhandler_test.cpp [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+ * Consts
+ **********************************************************************************************************************/
+
+/***********************************************************************************************************************
+ * Tests
+ **********************************************************************************************************************/
+
+TEST_F(PermHandlerTest, RegisterInstanceSucceeds)
+{
+    InstanceIdent                                instanceIdent;
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:52]: var definition should be separated with other operators. This function should look like:
```c++
TEST_F(PermHandlerTest, RegisterInstanceSucceeds)
{
    InstanceIdent                                instanceIdent;
    StaticArray<FunctionalServicePermissions, 1> perms;

    auto err = mPermHandler.RegisterInstance(instanceIdent, perms);
    EXPECT_TRUE(err.mError.IsNone()) << err.mError.Message();
    ASSERT_FALSE(err.mValue.IsEmpty());
}
```

File: tests/iam/permhandler/permhandler_test.cpp [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+
+    ASSERT_FALSE(err.mValue.IsEmpty());
+}
+
+TEST_F(PermHandlerTest, RegisterInstanceReturnsSecretFromCache)
+{
+    InstanceIdent                                instanceIdent;
+    StaticArray<FunctionalServicePermissions, 1> perms;
+
+    const auto lhsErr = mPermHandler.RegisterInstance(instanceIdent, perms);
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:66]: `auto err = `

File: tests/iam/permhandler/permhandler_test.cpp [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:67]: remove empty line

File: tests/iam/permhandler/permhandler_test.cpp [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+    EXPECT_TRUE(lhsErr.mError.IsNone()) << lhsErr.mError.Message();
+    ASSERT_FALSE(lhsErr.mValue.IsEmpty());
+
+    const auto rhsErr = mPermHandler.RegisterInstance(instanceIdent, perms);
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:71]: reuse err  above: `err = ...`

File: tests/iam/permhandler/permhandler_test.cpp [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+    EXPECT_EQ(lhsErr.mError, rhsErr.mError);
+    EXPECT_EQ(lhsErr.mValue, rhsErr.mValue);
+}
+
+TEST_F(PermHandlerTest, RegisterInstanceReachedMaxSize)
+{
+    for (size_t i {0}; i < cMaxNumInstances; ++i) {
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:78]: i = 0

File: tests/iam/permhandler/permhandler_test.cpp [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+        const InstanceIdent                          instanceIdent {"", "", i};
+        StaticArray<FunctionalServicePermissions, 1> perms;
+
+        const auto err = mPermHandler.RegisterInstance(instanceIdent, perms);
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:82]: `auto err`

File: tests/iam/permhandler/permhandler_test.cpp [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+        ASSERT_TRUE(err.mError.IsNone()) << err.mError.Message();
+        ASSERT_FALSE(err.mValue.IsEmpty());
+    }
+
+    const InstanceIdent                          instanceIdent {"", "", cMaxNumInstances};
+    StaticArray<FunctionalServicePermissions, 1> perms;
+
+    const auto err = mPermHandler.RegisterInstance(instanceIdent, perms);
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:90]: `auto err`

File: tests/iam/permhandler/permhandler_test.cpp [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+    ASSERT_TRUE(err.mError.Is(ErrorEnum::eNoMemory)) << err.mError.Message();
+    ASSERT_TRUE(err.mValue.IsEmpty());
+}
+
+TEST_F(PermHandlerTest, GetPermissionsNotFound)
+{
+    InstanceIdent                instanceIdent;
+    StaticArray<PermKeyValue, 1> perms;
+
+    const auto err = mPermHandler.GetPermissions(instanceIdent, perms, "uknownSecretUUID", "unknownServerId");
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:100]: `auto err`

File: tests/iam/permhandler/permhandler_test.cpp [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+
+    FunctionalServicePermissions testServicePermissions;
+    testServicePermissions.mName = "testService";
+    testServicePermissions.mPermissions.PushBack({"key1", "value1"});
+    testServicePermissions.mPermissions.PushBack({"key2", "value2"});
+
+    StaticArray<FunctionalServicePermissions, 2> funcServerPermissions;
+    funcServerPermissions.PushBack(testServicePermissions);
+
+    const auto errRegister = mPermHandler.RegisterInstance(instanceIdent1, funcServerPermissions);
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:116]: `auto registerResult = `

File: tests/iam/permhandler/permhandler_test.cpp [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+    ASSERT_TRUE(errRegister.mError.IsNone()) << errRegister.mError.Message();
+
+    const auto secret = errRegister.mValue;
+
+    InstanceIdent                resInstanceIdent;
+    StaticArray<PermKeyValue, 1> resServicePerms;
+
+    const auto errGetPermissions
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:124]: `auto permissionResult =`

File: tests/iam/permhandler/permhandler_test.cpp [hunk 29]
```diff
@@ -0,0 +1,211 @@
...
+    testServicePermissions.mPermissions.PushBack({"key1", "value1"});
+    testServicePermissions.mPermissions.PushBack({"key2", "value2"});
+
+    StaticArray<FunctionalServicePermissions, 2> funcServerPermissions;
+    funcServerPermissions.PushBack(testServicePermissions);
+
+    const auto errRegister = mPermHandler.RegisterInstance(instanceIdent1, funcServerPermissions);
+    ASSERT_TRUE(errRegister.mError.IsNone()) << errRegister.mError.Message();
+
+    const auto secret = errRegister.mValue;
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:119]: Why intermediate var is required?

File: tests/iam/permhandler/permhandler_test.cpp [hunk 30]
```diff
@@ -0,0 +1,211 @@
...
+{
+    InstanceIdent instanceIdent;
+    instanceIdent.mServiceID = "test-service-id";
+
+    StaticArray<FunctionalServicePermissions, 1> perms;
+
+    auto err = mPermHandler.UnregisterInstance(instanceIdent);
+    EXPECT_FALSE(err.IsNone()) << err.Message();
+
+    const auto errWithValue = mPermHandler.RegisterInstance(instanceIdent, perms);
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:140]: `auto registerResult`

File: tests/iam/permhandler/permhandler_test.cpp [hunk 31]
```diff
@@ -0,0 +1,211 @@
...
+    FunctionalServicePermissions systemCoreServicePermissions;
+    systemCoreServicePermissions.mName = "systemCore";
+    systemCoreServicePermissions.mPermissions.PushBack({"test1.*", "rw"});
+    systemCoreServicePermissions.mPermissions.PushBack({"test2", "r"});
+
+    StaticArray<FunctionalServicePermissions, 2> funcServerPermissions;
+    funcServerPermissions.PushBack(visServicePermissions);
+    funcServerPermissions.PushBack(systemCoreServicePermissions);
+
+    auto       errRegister = mPermHandler.RegisterInstance(instanceIdent1, funcServerPermissions);
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:170]: `registerResult`

File: tests/iam/permhandler/permhandler_test.cpp [hunk 32]
```diff
@@ -0,0 +1,211 @@
...
+    const auto secret1     = errRegister.mValue;
```
Comment (al1img) [inline tests/iam/permhandler/permhandler_test.cpp:171]: intermediate var is not required

File: src/iam/permhandler/permhandler.cpp (hunk 1 above)
Comment (mlohvynenko) [reply src/iam/permhandler/permhandler.cpp:34]: Done

File: include/aos/common/tools/log.hpp [hunk 33]
```diff
@@ -88,8 +89,8 @@ class LogModuleType {
 
     static const Array<const char* const> GetStrings()
     {
-        static const char* const sLogModuleTypeStrings[]
-            = {"default", "launcher", "servicemanager", "certhandler", "identhandler", "resourcemonitor", "pkcs11"};
+        static const char* const sLogModuleTypeStrings[] = {"default", "launcher", "servicemanager", "certhandler",
+            "identhandler", "permhandler", "resourcemonitor", "pkcs11"};
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply include/aos/common/tools/log.hpp]: done

File: tests/iam/permhandler/permhandler_test.cpp (hunk 29 above)
Comment (mlohvynenko) [reply tests/iam/permhandler/permhandler_test.cpp]: Refactored. Currently this UT has 2 secret variables (for 2 different instances) that are used in test

File: tests/iam/permhandler/permhandler_test.cpp (hunk 32 above)
Comment (mlohvynenko) [reply tests/iam/permhandler/permhandler_test.cpp]: this UT has 2 secret variables (for 2 different instances) that are used in test

File: src/iam/permhandler/permhandler.cpp [hunk 34]
```diff
@@ -0,0 +1,138 @@
...
+
+    for (const auto& it : result.mValue->mFuncServicePerms) {
+        if (it.mName == funcServerID) {
+            if (it.mPermissions.Size() > servicePermissions.MaxSize()) {
+                return AOS_ERROR_WRAP(ErrorEnum::eNoMemory);
+            }
+
+            servicePermissions = it.mPermissions;
+
+            return ErrorEnum::eNone;
```
Comment (MykolaSolyanko) [inline src/iam/permhandler/permhandler.cpp:82]: AOS_ERROR_WRAP ?

File: src/iam/permhandler/permhandler.cpp [hunk 35]
```diff
@@ -0,0 +1,138 @@
...
+        }
+    }
+
+    return AOS_ERROR_WRAP(ErrorEnum::eNotFound);
```
Comment (MykolaSolyanko) [inline src/iam/permhandler/permhandler.cpp:86]: in some methods you are using AOS_ERROR_WRAP in some not, I think should be consistent

Comment (mlohvynenko) [reply src/iam/permhandler/permhandler.cpp:86]: Applied your suggestion

File: src/iam/permhandler/permhandler.cpp (hunk 34 above)
Comment (mlohvynenko) [reply src/iam/permhandler/permhandler.cpp:82]: Applied your suggestion

**Review by MykolaSolyanko — APPROVED** [review]

**Review by mykola-kobets-epam — APPROVED** [review]

File: include/aos/iam/permhandler.hpp [hunk 36]
```diff
@@ -0,0 +1,127 @@
...
+    /**
+     * Retruns instance ident and permissions by secret UUID and functional server ID.
+     *
+     * @param[out] instanceIdent result instance ident.
+     * @param[out] servicePermissions result service permission.
+     * @param secretUUID secret UUID.
+     * @param funcServerID functional server ID.
+     * @returns Error.
+     */
+    Error GetPermissions(InstanceIdent& instanceIdent, Array<PermKeyValue>& servicePermissions,
```
Comment (mykola-kobets-epam) [inline include/aos/iam/permhandler.hpp:106]: first should be input parameters, output parameters should be in the end of parameter list.

File: include/aos/iam/permhandler.hpp [hunk 37]
```diff
@@ -0,0 +1,127 @@
...
+struct InstancePermissions {
+    StaticString<uuid::cUUIDStrLen>                            mSecretUUID;
+    InstanceIdent                                              mInstanceIdent;
+    StaticArray<FunctionalServicePermissions, cMaxNumServices> mFuncServicePerms;
+};
+
+/**
+ * Permission handler.
+ */
+class PermHandler {
```
Comment (mykola-kobets-epam) [inline include/aos/iam/permhandler.hpp:77]: Please add PermissionHandlerItf

**Review by mykola-kobets-epam — APPROVED** [review]

File: include/aos/iam/permhandler.hpp (hunk 37 above)
Comment (mlohvynenko) [reply include/aos/iam/permhandler.hpp]: Done

File: include/aos/iam/permhandler.hpp (hunk 36 above)
Comment (mlohvynenko) [reply include/aos/iam/permhandler.hpp]: Applied suggestion

**Pull request merged**